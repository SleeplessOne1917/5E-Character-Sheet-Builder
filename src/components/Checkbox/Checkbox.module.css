.checkbox {
    width: 1.5rem;
    height: 1.5rem;
    border: 2px solid var(--davy);
    background-color: var(--heaven-light);
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;
    border-radius: 0.5rem;
    margin-right: 0.5rem;
}

.checked {
    background-color: var(--highlight) !important;
}

.alternate.checked {
    background-color: var(--davy) !important;
}

.label {
    display: flex;
    align-items: center;
    font-size: 1.5rem;
    cursor: pointer;
    user-select: none;
    width: fit-content;
}

.alternate .check {
    color: var(--heaven-light);
}

.check {
    width: 1.2rem;
    height: 1.2rem;
    color: var(--davy);
    pointer-events: none;
}

.checkbox:focus-visible {
    outline: var(--focus-outline);
}

@media screen and (hover: hover) {
    .checkbox:hover {
        background-color: var(--highlight-light);
    }
}