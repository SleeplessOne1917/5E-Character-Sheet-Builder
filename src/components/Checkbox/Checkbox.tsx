'use client';

import { KeyboardEvent, useCallback, useState } from 'react';

import { CheckIcon } from '@heroicons/react/20/solid';
import classes from './Checkbox.module.css';
import { handleKeyDownEvent } from '../../services/handlerService';

type CheckboxProps = {
	checked?: boolean;
	label: string;
	onChange: (value: boolean) => void;
	useAlternateStyle?: boolean;
	hideLabel?: boolean;
	disabled?: boolean;
};

const Checkbox = ({
	checked,
	onChange,
	label,
	useAlternateStyle = false,
	hideLabel = false,
	disabled = false
}: CheckboxProps) => {
	const [isChecked, setIsChecked] = useState(checked ?? false);

	const handleChange = useCallback(() => {
		onChange(!isChecked);

		setIsChecked(prev => !prev);
	}, [onChange, setIsChecked, isChecked]);

	const handleKeyDown = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			handleKeyDownEvent(event, handleChange);
		},
		[handleChange]
	);

	return (
		<label
			onClick={!disabled ? handleChange : () => {}}
			className={`${classes.label}${disabled ? ` ${classes.disabled}` : ''}`}
			data-testid="checkbox"
		>
			<div
				className={`${classes.checkbox}${
					checked || isChecked ? ` ${classes.checked}` : ''
				}${useAlternateStyle ? ` ${classes.alternate}` : ''}${
					disabled ? ` ${classes.disabled}` : ''
				}`}
				tabIndex={disabled ? -1 : 0}
				role="checkbox"
				aria-disabled={disabled}
				aria-checked={isChecked}
				aria-label={label}
				onKeyDown={!disabled ? handleKeyDown : () => {}}
			>
				{(checked || isChecked) && <CheckIcon className={classes.check} />}
			</div>
			{!hideLabel && label}
		</label>
	);
};

export default Checkbox;
