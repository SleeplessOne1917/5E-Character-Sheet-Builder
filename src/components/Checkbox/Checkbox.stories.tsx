import { ComponentMeta, ComponentStory } from '@storybook/react';

import Checkbox from './Checkbox';

export default {
	title: 'Components/Checkbox',
	component: Checkbox
} as ComponentMeta<typeof Checkbox>;

const Template: ComponentStory<typeof Checkbox> = args => (
	<Checkbox {...args} />
);

export const Default = Template.bind({});

export const Checked = Template.bind({});
Checked.args = {
	checked: true
};

export const Alternate = Template.bind({});
Alternate.args = {
	checked: true,
	useAlternateStyle: true
};
