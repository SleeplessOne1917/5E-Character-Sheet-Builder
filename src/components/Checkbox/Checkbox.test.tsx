import * as stories from './Checkbox.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/testing-react';

const { Checked, Default, Alternate } = composeStories(stories);

describe('renders correctly', () => {
	it('by default', () => {
		render(<Default />);

		expect(screen.getByRole('checkbox')).toMatchSnapshot();
	});

	it('when checked', () => {
		render(<Checked />);

		expect(screen.getByRole('checkbox')).toMatchSnapshot();
	});

	it('when alternate', () => {
		render(<Alternate />);

		expect(screen.getByRole('checkbox')).toMatchSnapshot();
	});
});
