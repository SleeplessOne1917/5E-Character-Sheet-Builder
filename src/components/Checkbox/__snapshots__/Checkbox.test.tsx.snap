// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly by default 1`] = `
<label
  class="label"
  data-testid="checkbox"
>
  <div
    aria-checked="false"
    aria-disabled="false"
    class="checkbox"
    role="checkbox"
    tabindex="0"
  />
</label>
`;

exports[`renders correctly when alternate 1`] = `
<label
  class="label"
  data-testid="checkbox"
>
  <div
    aria-checked="true"
    aria-disabled="false"
    class="checkbox checked alternate"
    role="checkbox"
    tabindex="0"
  >
    <svg
      aria-hidden="true"
      class="check"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M16.704 4.153a.75.75 0 01.143 1.052l-8 10.5a.75.75 0 01-1.127.075l-4.5-4.5a.75.75 0 011.06-1.06l3.894 3.893 7.48-9.817a.75.75 0 011.05-.143z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</label>
`;

exports[`renders correctly when checked 1`] = `
<label
  class="label"
  data-testid="checkbox"
>
  <div
    aria-checked="true"
    aria-disabled="false"
    class="checkbox checked"
    role="checkbox"
    tabindex="0"
  >
    <svg
      aria-hidden="true"
      class="check"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M16.704 4.153a.75.75 0 01.143 1.052l-8 10.5a.75.75 0 01-1.127.075l-4.5-4.5a.75.75 0 011.06-1.06l3.894 3.893 7.48-9.817a.75.75 0 011.05-.143z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</label>
`;

exports[`renders correctly when disabled 1`] = `
<label
  class="label disabled"
  data-testid="checkbox"
>
  <div
    aria-checked="false"
    aria-disabled="true"
    class="checkbox disabled"
    role="checkbox"
    tabindex="-1"
  />
</label>
`;
